File: organic_grow/lib/core/controllers/cart_controller.dart
import 'package:flutter/material.dart';
import 'package:get/get.dart';
import 'package:organic_grow/config/app_color.dart';
import 'package:organic_grow/config/app_typography.dart';
import 'package:organic_grow/core/models/cart_item_model.dart';
import 'package:organic_grow/core/services/api_services.dart';

class CartController extends GetxController {
  var cartItems = <CartItem>[].obs;
  var totalAmount = 0.0.obs;
  var currentVendorId = ''.obs;
  var currentVendorName = ''.obs;
  var deliveryCharge = 0.0.obs;
  var isLoading = false.obs;

  @override
  void onInit() {
    super.onInit();
    fetchCartFromServer();
  }

  /// Fetch cart from server on app start
  Future<void> fetchCartFromServer() async {
    try {
      isLoading.value = true;
      final data = await ApiService.fetchCart();
      if (data != null && data['cart'] != null) {
        _parseServerCart(data['cart']);
      } else {
        _clearLocalState();
      }
    } catch (e) {
      debugPrint("Failed to fetch cart: $e");
    } finally {
      isLoading.value = false;
    }
  }

  /// Add item to cart via server API (handles vendor conflict)
  Future<void> addToCart(CartItem item) async {
    if (isLoading.value) return; // Guard against double-tap / double-call
    try {
      isLoading.value = true;
      final response = await ApiService.addToCart(item.id, quantity: item.quantity);

      if (response['success'] == true && response['cart'] != null) {
        // Successfully added
        _parseServerCart(response['cart']);
        Get.snackbar('Added! ✅', '${item.name} added to cart',
            snackPosition: SnackPosition.BOTTOM,
            duration: const Duration(seconds: 1));
      } else if (response['conflict'] == true) {
        // Vendor conflict — show Zomato-style dialog
        _showVendorSwitchDialog(item);
      }
    } catch (e) {
      Get.snackbar('Error', 'Failed to add to cart',
          snackPosition: SnackPosition.BOTTOM);
    } finally {
      isLoading.value = false;
    }
  }

  /// Replace cart on vendor switch (user confirmed)
  Future<void> _replaceCartWithItem(CartItem item) async {
    try {
      isLoading.value = true;
      final response = await ApiService.replaceCart(item.id, quantity: 1);
      if (response['success'] == true && response['cart'] != null) {
        _parseServerCart(response['cart']);
        Get.snackbar('Cart Updated! 🔄', 'Cart replaced with ${item.vendorName} items',
            snackPosition: SnackPosition.BOTTOM,
            duration: const Duration(seconds: 2));
      }
    } catch (e) {
      Get.snackbar('Error', 'Failed to replace cart',
          snackPosition: SnackPosition.BOTTOM);
    } finally {
      isLoading.value = false;
    }
  }

  /// Increase quantity of an item
  Future<void> increaseQuantity(String productId) async {
    try {
      final item = cartItems.firstWhere((e) => e.id == productId);
      final response = await ApiService.updateCartItem(productId, item.quantity + 1);
      if (response['success'] == true) {
        if (response['cart'] != null) {
          _parseServerCart(response['cart']);
        }
      }
    } catch (e) {
      debugPrint("Failed to increase quantity: $e");
    }
  }

  /// Decrease quantity of an item
  Future<void> decreaseQuantity(String productId) async {
    try {
      final item = cartItems.firstWhere((e) => e.id == productId);
      final newQty = item.quantity - 1;
      if (newQty <= 0) {
        await removeFromCart(productId);
        return;
      }
      final response = await ApiService.updateCartItem(productId, newQty);
      if (response['success'] == true) {
        if (response['cart'] != null) {
          _parseServerCart(response['cart']);
        } else {
          _clearLocalState();
        }
      }
    } catch (e) {
      debugPrint("Failed to decrease quantity: $e");
    }
  }

  /// Remove a specific item from cart
  Future<void> removeFromCart(String productId) async {
    try {
      final response = await ApiService.removeCartItem(productId);
      if (response['success'] == true) {
        if (response['cart'] != null) {
          _parseServerCart(response['cart']);
        } else {
          _clearLocalState();
        }
      }
    } catch (e) {
      debugPrint("Failed to remove from cart: $e");
    }
  }

  /// Clear entire cart
  Future<void> clearCart() async {
    try {
      await ApiService.clearCart();
      _clearLocalState();
    } catch (e) {
      debugPrint("Failed to clear cart: $e");
    }
  }

  /// Get quantity of a specific product in cart (for UI buttons)
  int getProductQuantity(String productId) {
    final idx = cartItems.indexWhere((e) => e.id == productId);
    if (idx >= 0) return cartItems[idx].quantity;
    return 0;
  }

  /// Check if a product is in cart
  bool isInCart(String productId) {
    return cartItems.any((e) => e.id == productId);
  }

  // ==========================================
  // PRIVATE HELPERS
  // ==========================================

  /// Parse server cart response into local state
  void _parseServerCart(Map<String, dynamic> cartJson) {
    final vendor = cartJson['vendorId'];
    if (vendor != null && vendor is Map<String, dynamic>) {
      currentVendorId.value = vendor['_id'] ?? '';
      currentVendorName.value = vendor['shopName'] ?? '';
      deliveryCharge.value = (vendor['deliveryCharge'] ?? 0).toDouble();
    }

    final List<dynamic> products = cartJson['products'] ?? [];
    cartItems.assignAll(products.map((p) {
      final prod = p['productId'];
      String name = '';
      String image = '';
      double price = (p['price'] ?? 0).toDouble();
      String unit = 'kg';

      if (prod != null && prod is Map<String, dynamic>) {
        name = prod['productName'] ?? '';
        final imgs = prod['images'] ?? [];
        if (imgs is List && imgs.isNotEmpty) {
          image = ApiService.buildImageUrl(imgs[0].toString().replaceAll('\\', '/'));
        }
        price = (prod['sellingPrice'] ?? prod['mrpPrice'] ?? price).toDouble();
        unit = prod['unit'] ?? 'kg';
      }

      return CartItem(
        id: (prod != null && prod is Map) ? (prod['_id'] ?? '') : (p['productId'] ?? ''),
        name: name,
        price: price,
        image: image,
        quantity: p['quantity'] ?? 1,
        vendorId: currentVendorId.value,
        vendorName: currentVendorName.value,
        unit: unit,
      );
    }).toList());

    totalAmount.value = (cartJson['totalPrice'] ?? 0).toDouble();
  }

  /// Clear all local cart state
  void _clearLocalState() {
    cartItems.clear();
    totalAmount.value = 0.0;
    currentVendorId.value = '';
    currentVendorName.value = '';
    deliveryCharge.value = 0.0;
  }

  /// Clear local cart state (public accessor for logout)
  void clearLocalCart() {
    _clearLocalState();
  }


  /// Zomato-style vendor switch confirmation dialog
  void _showVendorSwitchDialog(CartItem newItem) {
    Get.dialog(
      AlertDialog(
        shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(20)),
        title: Row(
          children: [
            const Icon(Icons.swap_horiz_rounded, color: Colors.orange, size: 28),
            const SizedBox(width: 8),
            Expanded(
              child: Text(
                'Replace cart items?',
                style: AppTypography.h4.copyWith(fontWeight: FontWeight.bold),
              ),
            ),
          ],
        ),
        content: Text(
          'Your cart contains items from "${currentVendorName.value}". '
          'Do you want to discard them and add items from "${newItem.vendorName}"?',
          style: AppTypography.bodyMedium.copyWith(
            color: AppColor.textColor.withOpacity(0.7),
          ),
        ),
        actions: [
          TextButton(
            onPressed: () => Get.back(),
            child: Text(
              'No',
              style: AppTypography.buttonMedium.copyWith(
                color: AppColor.textColor.withOpacity(0.5),
              ),
            ),
          ),
          ElevatedButton(
            onPressed: () {
              Get.back();
              _replaceCartWithItem(newItem);
            },
            style: ElevatedButton.styleFrom(
              backgroundColor: AppColor.primaryColor,
              foregroundColor: Colors.white,
              shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(12)),
            ),
            child: Text(
              'Yes, Replace',
              style: AppTypography.buttonMedium.copyWith(fontWeight: FontWeight.bold),
            ),
          ),
        ],
      ),
      barrierDismissible: true,
    );
  }
}

File: backend/controllers/cartController.js
const Cart = require("../models/Cart");
const Product = require("../models/Product");

// GET USER'S CURRENT CART
const getCart = async (req, res) => {
  try {
    const cart = await Cart.findOne({ userId: req.user.id })
      .populate("vendorId", "shopName shopImage deliveryTime deliveryCharge minimumOrder isOpen")
      .populate("products.productId", "productName images sellingPrice mrpPrice unit weight stock isAvailable");

    if (!cart) {
      return res.status(200).json({
        success: true,
        cart: null,
        message: "Cart is empty",
      });
    }

    res.status(200).json({
      success: true,
      cart,
    });
  } catch (error) {
    res.status(500).json({
      success: false,
      message: error.message,
    });
  }
};

// ADD ITEM TO CART (with single-vendor lock)
const addToCart = async (req, res) => {
  try {
    const { productId, quantity } = req.body;
    const userId = req.user.id;

    if (!productId) {
      return res.status(400).json({
        success: false,
        message: "productId is required",
      });
    }

    const qty = parseInt(quantity) || 1;

    // 1. Get the product to find its vendor
    const product = await Product.findById(productId);
    if (!product) {
      return res.status(404).json({
        success: false,
        message: "Product not found",
      });
    }

    if (!product.isAvailable) {
      return res.status(400).json({
        success: false,
        message: "Product is currently unavailable",
      });
    }

    // 2. Check if user already has a cart
    let cart = await Cart.findOne({ userId });

    if (!cart) {
      // CASE 1: No cart exists → create new cart with this vendor
      cart = await Cart.create({
        userId,
        vendorId: product.vendorId,
        products: [
          {
            productId: product._id,
            quantity: qty,
            price: product.sellingPrice || product.mrpPrice,
          },
        ],
      });

      // Recalculate total
      cart.totalPrice = cart.products.reduce(
        (sum, item) => sum + item.price * item.quantity,
        0
      );
      await cart.save();

      // Populate and return
      await cart.populate(
        "vendorId",
        "shopName shopImage deliveryTime deliveryCharge minimumOrder isOpen"
      );
      await cart.populate(
        "products.productId",
        "productName images sellingPrice mrpPrice unit weight stock isAvailable"
      );

      return res.status(201).json({
        success: true,
        message: "Item added to cart",
        cart,
      });
    }

    // CASE 3: Cart exists but DIFFERENT vendor → conflict!
    if (cart.vendorId.toString() !== product.vendorId.toString()) {
      return res.status(409).json({
        success: false,
        conflict: true,
        message:
          "Your cart has items from another store. Do you want to replace your cart?",
        currentVendorId: cart.vendorId,
        newVendorId: product.vendorId,
      });
    }

    // CASE 2: Same vendor → add or increment
    const existingItemIndex = cart.products.findIndex(
      (p) => p.productId.toString() === productId
    );

    if (existingItemIndex >= 0) {
      cart.products[existingItemIndex].quantity += qty;
    } else {
      cart.products.push({
        productId: product._id,
        quantity: qty,
        price: product.sellingPrice || product.mrpPrice,
      });
    }

    // Recalculate total
    cart.totalPrice = cart.products.reduce(
      (sum, item) => sum + item.price * item.quantity,
      0
    );
    await cart.save();

    await cart.populate(
      "vendorId",
      "shopName shopImage deliveryTime deliveryCharge minimumOrder isOpen"
    );
    await cart.populate(
      "products.productId",
      "productName images sellingPrice mrpPrice unit weight stock isAvailable"
    );

    res.status(200).json({
      success: true,
      message: "Item added to cart",
      cart,
    });
  } catch (error) {
    res.status(500).json({
      success: false,
      message: error.message,
    });
  }
};

// REPLACE CART (user confirmed vendor switch)
const replaceCart = async (req, res) => {
  try {
    const { productId, quantity } = req.body;
    const userId = req.user.id;

    const qty = parseInt(quantity) || 1;

    const product = await Product.findById(productId);
    if (!product) {
      return res.status(404).json({
        success: false,
        message: "Product not found",
      });
    }

    // Delete old cart and create new one with new vendor
    await Cart.deleteOne({ userId });

    const cart = await Cart.create({
      userId,
      vendorId: product.vendorId,
      products: [
        {
          productId: product._id,
          quantity: qty,
          price: product.sellingPrice || product.mrpPrice,
        },
      ],
      totalPrice: (product.sellingPrice || product.mrpPrice) * qty,
    });

    await cart.populate(
      "vendorId",
      "shopName shopImage deliveryTime deliveryCharge minimumOrder isOpen"
    );
    await cart.populate(
      "products.productId",
      "productName images sellingPrice mrpPrice unit weight stock isAvailable"
    );

    res.status(200).json({
      success: true,
      message: "Cart replaced with new store items",
      cart,
    });
  } catch (error) {
    res.status(500).json({
      success: false,
      message: error.message,
    });
  }
};

// UPDATE ITEM QUANTITY IN CART
const updateCartItem = async (req, res) => {
  try {
    const { productId, quantity } = req.body;
    const userId = req.user.id;

    const qty = parseInt(quantity);

    const cart = await Cart.findOne({ userId });
    if (!cart) {
      return res.status(404).json({
        success: false,
        message: "Cart not found",
      });
    }

    const itemIndex = cart.products.findIndex(
      (p) => p.productId.toString() === productId
    );

    if (itemIndex < 0) {
      return res.status(404).json({
        success: false,
        message: "Item not found in cart",
      });
    }

    if (qty <= 0) {
      // Remove item if quantity is 0 or less
      cart.products.splice(itemIndex, 1);

      // If cart is now empty, delete it
      if (cart.products.length === 0) {
        await Cart.deleteOne({ _id: cart._id });
        return res.status(200).json({
          success: true,
          message: "Cart is now empty",
          cart: null,
        });
      }
    } else {
      cart.products[itemIndex].quantity = qty;
    }

    // Recalculate total
    cart.totalPrice = cart.products.reduce(
      (sum, item) => sum + item.price * item.quantity,
      0
    );
    await cart.save();

    await cart.populate(
      "vendorId",
      "shopName shopImage deliveryTime deliveryCharge minimumOrder isOpen"
    );
    await cart.populate(
      "products.productId",
      "productName images sellingPrice mrpPrice unit weight stock isAvailable"
    );

    res.status(200).json({
      success: true,
      message: "Cart updated",
      cart,
    });
  } catch (error) {
    res.status(500).json({
      success: false,
      message: error.message,
    });
  }
};

// REMOVE SINGLE ITEM FROM CART
const removeCartItem = async (req, res) => {
  try {
    const { productId } = req.params;
    const userId = req.user.id;

    const cart = await Cart.findOne({ userId });
    if (!cart) {
      return res.status(404).json({
        success: false,
        message: "Cart not found",
      });
    }

    cart.products = cart.products.filter(
      (p) => p.productId.toString() !== productId
    );

    // If cart is now empty, delete it
    if (cart.products.length === 0) {
      await Cart.deleteOne({ _id: cart._id });
      return res.status(200).json({
        success: true,
        message: "Cart is now empty",
        cart: null,
      });
    }

    // Recalculate total
    cart.totalPrice = cart.products.reduce(
      (sum, item) => sum + item.price * item.quantity,
      0
    );
    await cart.save();

    await cart.populate(
      "vendorId",
      "shopName shopImage deliveryTime deliveryCharge minimumOrder isOpen"
    );
    await cart.populate(
      "products.productId",
      "productName images sellingPrice mrpPrice unit weight stock isAvailable"
    );

    res.status(200).json({
      success: true,
      message: "Item removed from cart",
      cart,
    });
  } catch (error) {
    res.status(500).json({
      success: false,
      message: error.message,
    });
  }
};

// CLEAR ENTIRE CART
const clearCart = async (req, res) => {
  try {
    await Cart.deleteOne({ userId: req.user.id });

    res.status(200).json({
      success: true,
      message: "Cart cleared",
      cart: null,
    });
  } catch (error) {
    res.status(500).json({
      success: false,
      message: error.message,
    });
  }
};

module.exports = {
  getCart,
  addToCart,
  replaceCart,
  updateCartItem,
  removeCartItem,
  clearCart,
};
